Dockerfile

```diff
@@ -1,80 +1,41 @@
-    rm -rf ./lib/* && \
-    git submodule sync && \
-    git submodule update --init --recursive --force
-
-COPY ./lib/picolibc-stdio.c /tinygo/lib/picolibc-stdio.c
+COPY ./Makefile /tinygo/Makefile
 
 RUN cd /tinygo/ && \
-    go install /tinygo/
+    make llvm-source
 
-# tinygo-wasm stage installs the needed dependencies to compile TinyGo programs for WASM.
-FROM tinygo-base AS tinygo-wasm
-
-COPY --from=tinygo-base /go/bin/tinygo /go/bin/tinygo
-COPY --from=tinygo-base /tinygo/src /tinygo/src
-COPY --from=tinygo-base /tinygo/targets /tinygo/targets
+# tinygo-llvm-build stage build the custom llvm with xtensa support
+FROM tinygo-llvm AS tinygo-llvm-build
 
 RUN cd /tinygo/ && \
-    apt-get update && \
-    apt-get install -y make clang-11 libllvm11 lld-11 cmake ninja-build && \
-    mkdir build && \
-    make wasi-libc binaryen
+    make llvm-build
 
-# tinygo-avr stage installs the needed dependencies to compile TinyGo programs for AVR microcontrollers.
-FROM tinygo-base AS tinygo-avr
+# tinygo-compiler stage builds the compiler itself
+FROM tinygo-llvm-build AS tinygo-compiler
 
-COPY --from=tinygo-base /go/bin/tinygo /go/bin/tinygo
-COPY --from=tinygo-base /tinygo/src /tinygo/src
-COPY --from=tinygo-base /tinygo/targets /tinygo/targets
-COPY --from=tinygo-base /tinygo/Makefile /tinygo/
-COPY --from=tinygo-base /tinygo/tools /tinygo/tools
-COPY --from=tinygo-base /tinygo/lib /tinygo/lib
+COPY . /tinygo
 
+# remove submodules directories and re-init them to fix any hard-coded paths
+# after copying the tinygo directory in the previous step.
 RUN cd /tinygo/ && \
-    apt-get update && \
-    apt-get install -y apt-utils make binutils-avr gcc-avr avr-libc && \
-    make gen-device-avr && \
-    apt-get autoremove -y && \
-    apt-get clean
-
-# tinygo-arm stage installs the needed dependencies to compile TinyGo programs for ARM microcontrollers.
-FROM tinygo-base AS tinygo-arm
-
-COPY --from=tinygo-base /go/bin/tinygo /go/bin/tinygo
-COPY --from=tinygo-base /tinygo/src /tinygo/src
-COPY --from=tinygo-base /tinygo/targets /tinygo/targets
-COPY --from=tinygo-base /tinygo/Makefile /tinygo/
-COPY --from=tinygo-base /tinygo/tools /tinygo/tools
-COPY --from=tinygo-base /tinygo/lib /tinygo/lib
+    rm -rf ./lib/* && \
+    git submodule sync && \
+    git submodule update --init --recursive --force
 
 RUN cd /tinygo/ && \
-    apt-get update && \
-    apt-get install -y apt-utils make clang-11 && \
-    make gen-device-nrf && make gen-device-stm32
+    make
 
-# tinygo-all stage installs the needed dependencies to compile TinyGo programs for all platforms.
-FROM tinygo-wasm AS tinygo-all
+# tinygo-tools stage installs the needed dependencies to compile TinyGo programs for all platforms.
+FROM tinygo-compiler AS tinygo-tools
 
-COPY --from=tinygo-base /tinygo/Makefile /tinygo/
-COPY --from=tinygo-base /tinygo/tools /tinygo/tools
-COPY --from=tinygo-base /tinygo/lib /tinygo/lib
+COPY --from=tinygo-compiler /tinygo/build/tinygo /go/bin/tinygo
```

Comment:
This looks redundant, because this stage already uses `FROM tinygo-compiler`

Just glancing over (sorry for the drive-by comment), but it looks like with this changes, the Dockerfile effectively become a single-stage build (with some intermediate "breakpoints"); was that the intent?

```dockerfile
FROM golang:1.17 AS tinygo-llvm
...
FROM tinygo-llvm AS tinygo-llvm-build
..
FROM tinygo-llvm-build AS tinygo-compiler
...
FROM tinygo-compiler AS tinygo-tools
```

From the above; `tinygo-tools` "extends" `tinygo-compiler`, "extends" `tinygo-llvm-build`, "extends" `tinygo-llvm`, "extends" `tinygo-llvm`, which "extends" `golang:1.17`

(admitted, before this, it looks like "most" things were already copied (`COPY --from=tinygo-base` etc))

This means that the final image will contain all sources that were copied in (and intermediate layers, some of which look to be deleting files)

Reply:
> This means that the final image will contain all sources that were copied in (and intermediate layers, some of which look to be deleting files)

Yes, the idea with the `dev` build was always that you could use it to add features to TinyGo and rebuild from within the image itself.

Reply:
Removed the redundant copy, thanks for the feedback @thaJeztah